ha/fault_tolerance/rest_api.py

```diff
@@ -0,0 +1,255 @@
+class Response:
+    pass
+
+class CcRestApi(ABC):
+
+    _app = None
+    _loop = None
+    _site = None
+    __is_shutting_down = False
+    _handle_signals = True
+    _signals = ('SIGINT', 'SIGTERM')
+
+    # Test function just to check whether CC REST API server is alive
+    # TODO: to be removed CORTX-30932
+    @staticmethod
+    async def handler_check_alive(request):
+        """
+        Test Handler function is added for testing purpose.
+        it returns test "CORTX CC REST API server is alive."
+        Args:
+            request (web.Request): Request object
+        Returns:
+            web.Response: response object that to be sent to client
+        """
+        return web.Response(text="CORTX CC REST API server is alive.")
+
+    @staticmethod
+    def init(handle_signals=True) -> None:
+        """
+        Initialize web application ans set supported routes, middlewares, etc.
+        Args:
+            handle_signals (bool, optional): Defaults to True.
+            If handle_signals is True then signal handler will be set.
+            if it is false then caller should handle the signals.
+            and needs to call CcRestApi.stop() before application terminates
+        """
+
+        CcRestApi._handle_signals = handle_signals
+
+        # Create Rest Application object and set middleware coroutine to it.
+        # A middleware is a coroutine that can modify either the request or response.
+        CcRestApi._app = web.Application(middlewares=[CcRestApi.rest_middleware])
+
+        # TODO: CORTX-30932 Add routs using view classes and remove below placeholder
+        # Note: Adding route '/' just to test
+        CcRestApi._app.router.add_get('/', CcRestApi.handler_check_alive)
+
+        CcRestApi._app.on_startup.append(CcRestApi._on_startup)
+        CcRestApi._app.on_shutdown.append(CcRestApi._on_shutdown)
+
+    @classmethod
+    async def check_for_unsupported_endpoint(cls, request):
+        """
+        Check unsupported endpoints.
+        Note: this function is expected to be call from middleware.
+        Args:
+            request (web.Request): Request object
+        """
+        # TODO: CORTX-30931 implement
+        pass
+
+    @staticmethod
+    def json_response(resp_obj, status=200):
+        """
+        Returns the the provided response object to client.
+        Args:
+            resp_obj (web.Response): Response object that to be sent to client.
+            status (int, optional): HTTP status code. Defaults to 200.
+        """
+        # TODO: CORTX-30931 Implement
+        pass
+
+    @staticmethod
+    def error_response(err: Exception, **kwargs):
+        """
+        Create response object from input exception object.
+        Args:
+            err (Exception): exception object for that error response sent to be client.
+        """
+        # TODO: CORTX-30931 Implement
+        pass
+
+    @staticmethod
+    @web.middleware
+    async def rest_middleware(request, handler):
```

Comment:
I see handler is registered at line no 56. But how these two args get passed to this?
Any internal knowledge of async I should have for this understanding?

Reply:
middleware are like the decorators in python.
They can add extra functionality to the responses, and then internally call handler func.
For better understanding, please refer to:
https://docs.aiohttp.org/en/stable/web_advanced.html#middlewares